key.core/src/main/java/de/uka/ilkd/key/rule/SetStatementRule.java

```diff
@@ -0,0 +1,135 @@
+import de.uka.ilkd.key.java.reference.ReferencePrefix;
+import de.uka.ilkd.key.java.statement.SetStatement;
+import de.uka.ilkd.key.logic.*;
+import de.uka.ilkd.key.logic.op.ProgramVariable;
+import de.uka.ilkd.key.logic.op.Transformer;
+import de.uka.ilkd.key.logic.op.UpdateApplication;
+import de.uka.ilkd.key.proof.Goal;
+
+import org.key_project.util.collection.ImmutableList;
+
+import org.jspecify.annotations.NonNull;
+
+/**
+ * A rule for set statements. This unwraps the contained CopyAssignment
+ *
+ * @author Julian Wiesler
+ */
+public final class SetStatementRule implements BuiltInRule {
+
+    /**
+     * The instance
+     */
+    public static final SetStatementRule INSTANCE = new SetStatementRule();
+    /**
+     * The name of this rule
+     */
+    private static final Name name = new Name("Set Statement");
+
+    private SetStatementRule() {
+        // no statements
+    }
+
+    @Override
+    public boolean isApplicable(Goal goal, PosInOccurrence occurrence) {
+        if (AbstractAuxiliaryContractRule.occursNotAtTopLevelInSuccedent(occurrence)) {
+            return false;
+        }
+        // abort if inside of transformer
+        if (Transformer.inTransformer(occurrence)) {
+            return false;
+        }
+
+        Term target = occurrence.subTerm();
+        if (target.op() instanceof UpdateApplication) {
+            target = UpdateApplication.getTarget(target);
+        }
+        final SourceElement activeStatement = JavaTools.getActiveStatement(target.javaBlock());
+        return activeStatement instanceof SetStatement;
+    }
+
+    @Override
+    public boolean isApplicableOnSubTerms() {
+        return false;
+    }
+
+    @Override
+    public IBuiltInRuleApp createApp(PosInOccurrence occurrence, TermServices services) {
+        return new SetStatementBuiltInRuleApp(this, occurrence);
+    }
+
+    @NonNull
+    @Override
+    public ImmutableList<Goal> apply(Goal goal, Services services, RuleApp ruleApp)
+            throws RuleAbortException {
+        if (!(ruleApp instanceof SetStatementBuiltInRuleApp)) {
+            throw new IllegalArgumentException("can only apply SetStatementBuiltInRuleApp");
+        }
+
+        final TermBuilder tb = services.getTermBuilder();
+        final PosInOccurrence occurrence = ruleApp.posInOccurrence();
+        final Term formula = occurrence.subTerm();
+        assert formula.op() instanceof UpdateApplication
+                : "Currently, this can only be applied if there is an update application in front of the modality";
+
+        Term update = UpdateApplication.getUpdate(formula);
+        Term target = UpdateApplication.getTarget(formula);
+
+        SetStatement setStatement =
+            Optional.ofNullable(JavaTools.getActiveStatement(target.javaBlock()))
+                    .filter(SetStatement.class::isInstance).map(SetStatement.class::cast)
+                    .orElseThrow(() -> new RuleAbortException("not a JML set statement."));
```

Comment:
just a comment: I don't know, but this functional style does not look so convining in Java ...

Reply:
I know. A reason I use Kotlin. This code is from Julian. 

I did not add any functionality, I just moved it around.